### DISCUSSION TITLE
Please detail how to use one server to run multiple apps on different ports

### DISCUSSION THREAD

**Discussion by mariusa** [piku/piku#286] [General]
Hi, just discovered Piku!

> Bwahahahaha. There are no containers here, my friend. There are no orchestrators.

I'm trying to migrate away from Heroku and all howtos explain how to setup a container for each app, db, redis etc. and even how hard it is to run a cron job on docker for workers.
Thank you for a great solution which keeps it simple!

Question: Would you please detail how to use one server to run multiple apps on different ports, and how exactly nginx works with all this? In the end, will I have 
server.com:8001 for app1
server.com:8002 for app3
server.com:8003 for app2

and users have to enter port in the web browser to use them? Or can one simply have
https://app1.server.com
https://app2.server.com
https://app3.server.com

(all on the same server)

Thank you!

**Comment by rcarmo** [comment]
You can have both - however, we only use `nginx` for hostnames - in a typical Linux box, and assuming you're using, say, a Python or Node app, you just need to bind `web` in `app1` to `0.0.0.0:8001` and your little daemon will listen on that port - directly, without `nginx`.

The recommended way to put _anything_ on the Internet is to use `nginx` as a front-end/reverse proxy with `NGINX_SERVER_NAME` set in `ENV`. Right now using `nginx` with a custom port name has to be done with `NGINX_INCLUDE_FILE` (see [`ENV.md`](https://github.com/piku/piku/blob/master/docs/ENV.md) for more options).

**Comment by chr15m** [comment]
@mariusa and also anybody who finds this in the future, please note the easiest way is to have your web server bind to the environment variable called `PORT`. If you do that and follow what @rcarmo said, you will have the simplest experience as Nginx will automatically link the domain to to the correct server.

**Reply by kaseyhinton** [reply]
I don't know if it's helpful but examples that I use for configuring ports for my projects look like this


### Example 1
You can pass in the port command line style the --port being specific for whatever server you are using (parcel, http-server, servor, or equivalent)

// Procfile
`web: parcel index.html --port=$PORT`


### Example 2

You can set the port in the application code (this example is express)

// Procfile
`web: node server.js`

// server.js
```
var port = process.env["PORT"];
app.listen(port, () => {
  console.log("server started on port", port);
});
```

I should note that I am not determining the discrete port to use on my side. Piku gives me the port. All I have to do is have my application code grab the proper port from the environment.
